bchec/ecmh.go

```diff
@@ -0,0 +1,91 @@
+package bchec
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"github.com/gcash/bchd/chaincfg/chainhash"
+	"math/big"
+)
+
+// Multiset tracks the state of a multiset as used to calculate the ECMH
+// (elliptic curve multiset hash) hash of an unordered set. The state is
+// a point on the curve. New elements are hashed onto a point on the curve
+// and then added to the current state. Hence elements can be added in any
+// order and we can also remove elements to return to a prior hash.
+type Multiset struct {
+	curve *KoblitzCurve
+	x     *big.Int
+	y     *big.Int
+}
+
+// NewMultiset returns an empty multiset. The hash of an empty set
+// is the 32 byte value of zero.
+func NewMultiset(curve *KoblitzCurve) *Multiset {
+	return &Multiset{curve: curve}
+}
+
+// Add hashes the data onto the curve and updates the state
+// of the multiset.
+func (ms *Multiset) Add(data []byte) {
+	x, y := hashToPoint(ms.curve, data)
+	if ms.x == nil || ms.y == nil {
+		ms.x, ms.y = x, y
+	} else {
+		ms.x, ms.y = ms.curve.Add(ms.x, ms.y, x, y)
+	}
+}
+
+// Remove hashes the data onto the curve and subtracts the value
+// from the state. This function will execute regardless of whether
+// or not the passed data was previously added to the set. Hence,
+// adding a bunch of elements and then removing them all will not
+// result in a zero hash.
+func (ms *Multiset) Remove(data []byte) {
+	x, y := hashToPoint(ms.curve, data)
+	fx1, fy1 := ms.curve.bigAffineToField(x, y)
+	fy1 = fy1.Negate(1)
```

Comment:
I don't know if using 1 for the magnitude here is always correct. It passes the test vectors but is this value always correct?

Reply:
Created a multiset with millions of elements and removing each one still returned the correct hash. That would seem to suggest 1 is the correct value.

Reply:
Yes, I would say this is correct.